core/domain/story_services.py

```diff
@@ -111,6 +112,11 @@ def apply_change_list(story_id, change_list):
             if change.cmd == story_domain.CMD_ADD_STORY_NODE:
                 story.add_node(change.node_id, change.title)
             elif change.cmd == story_domain.CMD_DELETE_STORY_NODE:
+                for node in story.story_contents.nodes:
+                    if node.id == change.node_id:
+                        exp_services.update_exploration_story_link(
```

Comment:
OK, one concern here -- the rest of this is updating the domain objects, but here you are updating storage models. That should not be the case. You might need to update the signature of this function so that whatever is returned includes info about context models that need to be changed, and then only apply the model changes at the same time as the rest of the changes to other models.

Reply:
Done, added 2 arrays for expIds to add and to remove (with proper logic to prevent overlap) and used that at the end to do everything simultaneously.